runtime/runc_jailer.go

```diff
@@ -0,0 +1,405 @@
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev, uid, gid int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, rdev); err != nil {
+		return err
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+func (j runcJailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.CommandContext(j.ctx, j.runcBinaryPath, "run", containerName)
+	cmd.Dir = j.OCIBundlePath()
+	return cmd
+}
+
+// overwriteConfig will set the proper default values if a field had not been set.
+//
+// TODO: Add netns
+func (j runcJailer) overwriteConfig(cfg *Config, socketPath, configPath string) error {
+	spec := specs.Spec{}
+	configBytes, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(configBytes, &spec); err != nil {
+		return err
+	}
+
+	if spec.Process.User.UID != 0 ||
+		spec.Process.User.GID != 0 {
+		return fmt.Errorf(
+			"using UID %d and GID %d, these values must not be set",
+			spec.Process.User.UID,
+			spec.Process.User.GID,
+		)
+	}
+
+	spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+
+	spec.Root.Path = rootfsFolder
+	spec.Root.Readonly = false
+	spec.Process.User.UID = j.uid
+	spec.Process.User.GID = j.gid
+
+	configBytes, err = json.Marshal(&spec)
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(configPath, configBytes, 0444); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setDefaultConfigValues will process the spec file provided and allow any
+// empty/zero values to be replaced with default values.
+func (j runcJailer) setDefaultConfigValues(cfg *Config, socketPath string, spec specs.Spec) specs.Spec {
+	if spec.Process == nil {
+		spec.Process = &specs.Process{}
+	}
+
+	if spec.Process.Args == nil {
+		cmd := firecracker.VMCommandBuilder{}.
+			WithBin("/firecracker").
+			WithSocketPath(socketPath).
+			// Don't need to pass in an actual context here as we are only building
+			// the command arguments and not actually building a command
+			Build(context.Background())
+
+		spec.Process.Args = cmd.Args
+	}
+
+	return spec
+}
+
+func mkdirAndChown(path string, mode os.FileMode, uid, gid uint32) error {
+	if err := os.MkdirAll(path, mode); err != nil {
```

Comment:
How about using Mkdir() instead of MkdirAll()? MkdirAll can create /foo/bar/baz even there is no /fo, but Chown only modifies /baz in this case.

Reply:
Fixed